arcgis-ios-sdk-samples/Geometry/Buffer list/BufferListViewController.swift

```diff
@@ -0,0 +1,238 @@
+        didSet {
+            undoBarButtonItem.isEnabled = !tappedPointsAndRadius.isEmpty
+            clearBarButtonItem.isEnabled = !tappedPointsAndRadius.isEmpty
+            // Redraw the buffers.
+            drawBuffers()
+        }
+    }
+    
+    /// The radius of the buffer.
+    var bufferRadius: Measurement<UnitLength> = Measurement(value: 0, unit: .miles)
+    /// A formatter for the output distance string.
+    let distanceFormatter: MeasurementFormatter = {
+        let formatter = MeasurementFormatter()
+        formatter.numberFormatter.maximumFractionDigits = 0
+        return formatter
+    }()
+    
+    // MARK: Instance methods
+    
+    /// Creates a map with an image base layer.
+    ///
+    /// - Parameter imageLayer: An `AGSArcGISMapImageLayer` object as the base layer of the map.
+    /// - Returns: A new `AGSMap` object.
+    func makeMap(imageLayer: AGSArcGISMapImageLayer) -> AGSMap {
+        // The spatial reference for this sample.
+        let statePlaneNorthCentralTexas = AGSSpatialReference(wkid: 32038)!
+        let map = AGSMap(spatialReference: statePlaneNorthCentralTexas)
+        map.basemap.baseLayers.add(imageLayer)
+        return map
+    }
+    
+    // MARK: Actions
+    
+    @IBAction func undoButtonTapped(_ sender: UIBarButtonItem) {
+        // Remove the last pair of tapped point and radius.
+        _ = tappedPointsAndRadius.popLast()
+    }
+    
+    @IBAction func isUnionSwitchValueChanged(_ sender: UISwitch) {
+        // Redraw buffers when the union switch's value is changed.
+        drawBuffers()
+    }
+    
+    @IBAction func clearButtonTapped(_ sender: UIBarButtonItem) {
+        bufferGraphicsOverlay.graphics.removeAllObjects()
+        tappedLocationsGraphicsOverlay.graphics.removeAllObjects()
+        tappedPointsAndRadius.removeAll()
+    }
+    
+    // MARK: UI
+    
+    func setStatus(message: String) {
+        statusLabel.text = message
+    }
+    
+    func drawBuffers() {
+        // Clear existing buffers graphics before drawing.
+        bufferGraphicsOverlay.graphics.removeAllObjects()
+        tappedLocationsGraphicsOverlay.graphics.removeAllObjects()
+        
+        guard !tappedPointsAndRadius.isEmpty else {
+            setStatus(message: "Tap on the map to add buffers.")
+            return
+        }
+        
+        // Create the buffers.
+        // Notice: the radius distances has the same unit of the map's spatial reference's unit.
+        // In this case, the statePlaneNorthCentralTexas spatial reference uses US feet.
+        if let bufferPolygon = AGSGeometryEngine.bufferGeometries(tappedPointsAndRadius.map { $0.point }, distances: tappedPointsAndRadius.map { NSNumber(value: $0.radius) }, unionResults: isUnionSwitch.isOn) {
+            // Add graphics symbolizing the tap point.
+            tappedLocationsGraphicsOverlay.graphics.addObjects(from: tappedPointsAndRadius.map { AGSGraphic(geometry: $0.point, symbol: nil) })
```

Comment:
Instead of iterating over `tappedPointsAndRadius` multiple times, I would suggest iterating over it once to get the array of points and radii:

```suggestion
        let (points, radii) = tappedPointsAndRadius.reduce(into: ([AGSPoint](), [NSNumber]())) { (result, pointAndRadius) in
            result.0.append(pointAndRadius.point)
            result.1.append(pointAndRadius.radius as NSNumber)
        }
        if let bufferPolygon = AGSGeometryEngine.bufferGeometries(points, distances: radii, unionResults: isUnionSwitch.isOn) {
            // Add graphics symbolizing the tap point.
            tappedLocationsGraphicsOverlay.graphics.addObjects(from: points.map { AGSGraphic(geometry: $0, symbol: nil) })
```

FYI: Double can be unconditionally cast to `NSNumber` (`double as NSNumber`, `doubles as [NSNumber]`, etc.). I don't know if casting is more efficient that initializing, but it is the preferred approach.